File: python/binary_tree_paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {string[]}
    def binaryTreePaths(self, root):
        if root is None:
            return []
        self.allPaths = []
        
        self.travel(root, "")
            
        return self.allPaths
    
    def travel(self, root, path):
        # Is leaf.
        if root.left is None and root.right is None:
            self.allPaths.append(path + str(root.val))
            return
        if root.left is not None:
            self.travel(root.left, path + str(root.val) + "->")
        if root.right is not None:
            self.travel(root.right, path + str(root.val) + "->")

File: cpp/binary_tree_paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector<string> allPaths;
    
    void travel(TreeNode* root, const std::string &path) {
        ostringstream oss;
        // is leaf
        if (root->left == nullptr && root->right == nullptr) {
            oss << path << root->val;
            allPaths.push_back(oss.str());
            return;
        }
        oss << path << root->val << "->";
        
        if (root->left != nullptr) 
            travel (root->left, oss.str());
        if (root->right != nullptr)
            travel (root->right, oss.str());
    }
    
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        if (root == nullptr)
            return allPaths;
        
        travel(root, "");
        
        return allPaths;
    }
};